client-mr/mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/CelebornShuffleConsumer.java

```diff
@@ -0,0 +1,173 @@
+import org.apache.celeborn.util.HadoopUtils;
+
+public class CelebornShuffleConsumer<K, V>
+    implements ShuffleConsumerPlugin<K, V>, ExceptionReporter {
+  private static final Logger logger = LoggerFactory.getLogger(CelebornShuffleConsumer.class);
+  private JobConf mrJobConf;
+  private MergeManager<K, V> merger;
+  private Throwable throwable = null;
+  private Progress copyPhase;
+  private TaskStatus taskStatus;
+  private org.apache.hadoop.mapreduce.TaskAttemptID reduceId;
+  private TaskUmbilicalProtocol umbilical;
+  private Reporter reporter;
+  private ShuffleClientMetrics metrics;
+  private Task reduceTask;
+
+  private ShuffleClient shuffleClient;
+
+  @Override
+  public void init(Context<K, V> context) {
+
+    reduceId = context.getReduceId();
+    mrJobConf = context.getJobConf();
+    JobConf celebornJobConf = new JobConf(HadoopUtils.MR_CELEBORN_CONF);
+
+    umbilical = context.getUmbilical();
+    reporter = context.getReporter();
+    try {
+      this.metrics = createMetrics(reduceId, mrJobConf);
+    } catch (Exception e) {
+      logger.error("Fatal error occurred, failed to get shuffle client metrics.", e);
+      reportException(e);
+    }
+    copyPhase = context.getCopyPhase();
+    taskStatus = context.getStatus();
+    reduceTask = context.getReduceTask();
+
+    String appId = celebornJobConf.get(HadoopUtils.MR_CELEBORN_APPLICATION_ID);
+    String lcHost = celebornJobConf.get(HadoopUtils.MR_CELEBORN_LC_HOST);
+    int lcPort = Integer.parseInt(celebornJobConf.get(HadoopUtils.MR_CELEBORN_LC_PORT));
+    logger.info("Reducer initialized with celeborn {} {} {}", appId, lcHost, lcPort);
+    CelebornConf celebornConf = HadoopUtils.fromYarnConf(mrJobConf);
+    shuffleClient =
+        ShuffleClient.get(
+            appId,
+            lcHost,
+            lcPort,
+            celebornConf,
+            new UserIdentifier(
+                celebornConf.quotaUserSpecificTenant(), celebornConf.quotaUserSpecificUserName()));
+    this.merger = createMergeManager(context);
+  }
+
+  // Merge mapOutput and spill in local disks if necessary
+  protected MergeManager<K, V> createMergeManager(ShuffleConsumerPlugin.Context context) {
+    return new MergeManagerImpl<K, V>(
+        reduceId,
+        mrJobConf,
+        context.getLocalFS(),
+        context.getLocalDirAllocator(),
+        reporter,
+        context.getCodec(),
+        context.getCombinerClass(),
+        context.getCombineCollector(),
+        context.getSpilledRecordsCounter(),
+        context.getReduceCombineInputCounter(),
+        context.getMergedMapOutputsCounter(),
+        this,
+        context.getMergePhase(),
+        context.getMapOutputFile());
+  }
+
+  private ShuffleClientMetrics createMetrics(
+      org.apache.hadoop.mapreduce.TaskAttemptID taskAttemptID, JobConf jobConf)
+      throws NoSuchMethodException, InvocationTargetException, InstantiationException,
+          IllegalAccessException {
+    // for hadoop 3
```

Comment:
we can use `DynMethods` to improve code readability, like what we did in the class `MemoryManager`.

https://github.com/apache/incubator-celeborn/blob/e2c42faed9b268de38420ed94fe64a3b09a940d4/worker/src/main/java/org/apache/celeborn/service/deploy/worker/memory/MemoryManager.java#L115-L120

Reply:
Sounds good to me. I'll try it.